### PR TITLE
fix: added deploying hydra to production guide

### PR DESCRIPTION
**Pull request by gen1us2k**
## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md)
      and signed the CLA.
- [x] I have referenced an issue containing the design document if my change
      introduces a new feature.
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security vulnerability, I
      confirm that I got green light (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [x] I have added tests that prove my fix is effective or that my feature
      works.
- [x] I have added necessary documentation within the code base (if
      appropriate).

### CODE REVIEW

**Comment by aeneasr** [comment]
Andrew, this is grand! @tomekpapiernik @vinckr when reviewing can you please follow the steps 1:1 to ensure we did not miss anything? :)

**Comment by FredLackeyOfficial** [comment]
Definite improvement on the current docs.  The only negative is it tends to answer the HOW question and not the WHY.  For example, the concept of an admin port.  What is that thing?  Is it for back-end administrative calls between services?  If so, then why would anyone ever make it public.  If it's for use by a Hydra administrator, then... well... the same question... why make it public?

**Comment by vinckr** [comment]
Hey @FredLackeyOfficial 

Would this document answer your questions why Hydra has an admin/public port: 
https://www.ory.sh/docs/hydra/production#exposing-administrative-and-public-api-endpoints
?

**Comment by FredLackeyOfficial** [comment]
@vinckr Yes and no.  Yes, the document you listed does help explain.  Unfortunately, by the time someone finds that doc it's too late.  Since every other example in the Ory site shows the admin port being exposed, the reader is left with the belief this is as a normal / acceptable practice for Hydra.  And, since they see "admin" being used publicly then they begin to question what that port REALLY is... since nobody would ever expose admin calls publicly.  My assumption is that is why this PR also follows the same pattern.  

When people come to the Ory site, they know nothing of the product or how its to be used.  Abbreviated examples set them on their way without any real explanation.  The end result is people just think they MUST continue the pattern that worked when they first ran the tutorial... which, in this example, almost always includes a public admin port.

This contributor's example does help by removing the "black box" haze caused by wrapping demos in docker-compose and the like.  My only point of commenting was to point out that, even with this improved example, the immediate need by Ory is to to explain WHY certain practices are needed, and HOW it functions.  The docs currently offer two vastly different extremes: either far too simple, with no explanations, or theoretical discussions... again, with no real examples.

**Comment by vinckr** [comment]
Thanks Fred, lets continue the discussion in the issue, I left a comment: https://github.com/ory/docs/issues/729

File: docs/hydra/guides/digital-ocean.mdx [hunk 1]
```diff
@@ -0,0 +1,465 @@
...
+SSL connection (protocol: TLSv1.3, cipher: TLS_AES_256_GCM_SHA384, bits: 256, compression: off)
+Type "help" for help.
+```
+
+We installed and configured the PostgreSQL database. 
+Next we will set up Ory Hydra.
+
+## Install Ory Hydra
+
+For security reasons, we need to create a new user with prohibited login:
```
Comment (vinckr) [inline docs/hydra/guides/digital-ocean.mdx:164]: elaborate, 
what are we securing against?

File: docs/hydra/guides/digital-ocean.mdx [hunk 2]
```diff
@@ -0,0 +1,465 @@
...
+  'hydra'
+)}/contrib/quickstart/5-min/hydra.yml`}</CodeBlock>
+
+We need to add some configuration in `hydra.yml` to use our Postgres
+
+```yml
+dsn: postgres://hydra:b0qw68gr3Q@127.0.0.1:5432/hydra?sslmode=disable&max_conns=20&max_idle_conns=4
+urls:
+  self:
+    issuer: https://accounts.gen1us2k.com
```
Comment (vinckr) [inline docs/hydra/guides/digital-ocean.mdx:202]: what is this part for? would be good to explain it

the DSN setting we can also mention, to connect the database.

**Review by vinckr — CHANGES_REQUESTED** [review]
I followed the tutorial until the end and everything worked quite well. 

I think it needs some steps in the end to test that Hydra is running correctly (/health check?).
Also some next steps would be helpful if can think of any from the top of your mind.

Overall it could use a bit more explanation in a few places, we just need to take care not to duplicate content that exists elsewhere. Ideally we link to an existing document - I left some comments where we should add some more detail IMO. 

Whenever you use `Let's` you can also leave it out and directly address the reader. (Instead of "Lets check our XY" -> "Check your XY")

**Comment by gen1us2k** [comment]
@FredLackeyOfficial Hello. Thanks for your feedback, and I added a small description of locations configuration with an explanation.

**Review by vinckr — APPROVED** [review]

**Review by aeneasr — DISMISSED** [review]
Before merging this, please address my feedback from the other PR on Kratos. We need to probably make those changes here too!

Also, would it make sense to put "Create a Droplet" into a common `mdx` file which we import?

**Comment by vinckr** [comment]
I agree, we can first merge the Kratos guide and then use it as template for this guide as large parts are similar. 
Marking this as draft until we merge the Kratos guide and then we give it a onceover and merge this one. 

I actually followed the Hydra deploy guide to the end without any problems, so I guess we mostly need to add more links to other places in the docs and "next steps" and so on, after we ensure its the same structure as the kratos guide.

**Comment by vinckr** [comment]
Added some next steps and made sure the layout and everything else is like in the Kratos guide.

**Comment by vinckr** [comment]
fixing a few broken links

**Review by vinckr — APPROVED** [review]
LGTM 
I ran the tutorial and had a working oauth2.vinckr.com instance afterwards (droplet is already destroyed now)

File: docs/hydra/guides/deploy-hydra-example.mdx [hunk 3]
```diff
@@ -0,0 +1,494 @@
...
+tar xfvz hydra_${useLatestRelease('hydra')}-linux_64bit.tar.gz
+# Remove redundant files
+rm *md
+rm LICENSE
+`}</CodeBlock>
+
+Download the [Quickstart](../5min-tutorial) configuration files:
+
+<CodeBlock className="language-shell">{`cd ../config
+wget https://raw.githubusercontent.com/ory/hydra/${useLatestRelease('hydra')}/contrib/quickstart/5-min/hydra.yml`}</CodeBlock>
```
Comment (tomekpapiernik) [inline docs/hydra/guides/deploy-hydra-example.mdx:166]: I think it'd be better to replace these code blocks with regular markdown code blocks aka "the three ticks".

File: docs/hydra/guides/deploy-hydra-example.mdx [hunk 4]
```diff
@@ -0,0 +1,494 @@
...
+
+We need to add some configuration in `hydra.yml` to use our Postgres We need configure Ory Hydra to use the Postgres database.
```
Comment (tomekpapiernik) [inline docs/hydra/guides/deploy-hydra-example.mdx:168]:
```suggestion
We need to add some configuration in `hydra.yml` to use our Postgres We need configure Ory Hydra to use the Postgres database.
```

Something went wrong here, not sure what this sentence was supposed to sound like?

File: docs/hydra/guides/deploy-hydra-example.mdx [hunk 5]
```diff
@@ -0,0 +1,494 @@
...
+the scale of your application.
+
+:::
+
+Wait for the virtual machine (VM) to start up and copy the IP address.
+
+This guide uses `oauth2.example.com` as a hostname to run Ory Hydra. Replace it with `<something>.<your-domain>` for your
+purposes. You need to configure DNS, create an `A type` record, and point it to the VM's IP.
+
+## Install required dependencies
```
Comment (tomekpapiernik) [inline docs/hydra/guides/deploy-hydra-example.mdx:46]: It'd make sense to turn this paragraph into an ordered list to show the flow better.

File: docs/hydra/guides/deploy-hydra-example.mdx [hunk 6]
```diff
@@ -0,0 +1,494 @@
...
+psql (12.9 (Ubuntu 12.9-0ubuntu0.20.04.1))
+SSL connection (protocol: TLSv1.3, cipher: TLS_AES_256_GCM_SHA384, bits: 256, compression: off)
+Type "help" for help.
+
+hydra=>
+```
+
+Congratulations, you have installed and configured the PostgreSQL database. Next we will set up Ory Hydra.
+
+## Install Ory Hydra
```
Comment (tomekpapiernik) [inline docs/hydra/guides/deploy-hydra-example.mdx:134]: It'd make sense to turn this paragraph into an ordered list to show the flow better.

File: docs/hydra/guides/deploy-hydra-example.mdx [hunk 7]
```diff
@@ -0,0 +1,494 @@
...
+Take security seriously and subscribe to the Ory Security Newsletter. Stay on top of new patches and security insights.
+
+>> Subscribe now: http://eepurl.com/di390P <<
+INFO[2022-04-15T15:24:37Z] Software quality assurance features are enabled. Learn more at: https://www.ory.sh/docs/ecosystem/sqa  audience=application service_name=Ory Hydra service_version=v1.11.7
+WARN[2022-04-15T15:24:37Z] JSON Web Key Set "hydra.https-tls" does not exist yet, generating new key pair...  audience=application service_name=Ory Hydra service_version=v1.11.7
+INFO[2022-04-15T15:24:41Z] Setting up http server on :4444               audience=application service_name=Ory Hydra service_version=v1.11.7
+INFO[2022-04-15T15:24:41Z] Setting up http server on :4445               audience=application service_name=Ory Hydra service_version=v1.11.7
+```
+
+## Run Ory Hydra using systemd
```
Comment (tomekpapiernik) [inline docs/hydra/guides/deploy-hydra-example.mdx:208]: It'd make sense to turn this paragraph into an ordered list to show the flow better.

**Review by tomekpapiernik — CHANGES_REQUESTED** [review]
There's a lot of work to be done before this guide gets published, at least from the clarity/language pov.

**Comment by tomekpapiernik** [comment]
If this guide is as we want from the technical point of view and we want to add it to the Ory documentation, then someone has to take care of the structure / language of this document.

Let's establish if we want to publish this first. If the answer is "yes", then let's decide who will on polishing it ;)

**Comment by vinckr** [comment]
Thanks for the thorough review @tomekpapiernik .

Please note that we merged basically the same guide for Kratos already.

https://www.ory.sh/docs/kratos/guides/deploy-kratos-example

I would propose to address your comments here in one fell swoop and then also make the same/similar changes in the Kratos doc, wdyt?

Btw this guide is not intended for beginners (that is the quickstart), so we expect some knowledge of VMs, linux and so on.

**Comment by tomekpapiernik** [comment]
Sounds good! Most of the things I commented on I can take care on my own. For a few questions I'm going to need your help. I'm going to tag you so you can see where exactly I need your input.

File: docs/hydra/guides/deploy-hydra-example.mdx [hunk 8]
```diff
@@ -0,0 +1,494 @@
...
+- **CPU options**: Regular with SSD
+- **RAM**: 1Gb
+- **SSD**: 25Gb
+- **VPC network**: default
+- **Authentication**: SSH Keys. Don't forget to add your own SSH key
+- **Region**: Choose your region
+
+:::note
+
+This example shows a single configuration on a small droplet. The configuration of a virtual machine (VM) may vary depending on
```
[the remark this answers is not in the document]
Comment (tomekpapiernik) [reply docs/hydra/guides/deploy-hydra-example.mdx:36]: @vinckr __

File: docs/hydra/guides/deploy-hydra-example.mdx [hunk 9]
```diff
@@ -0,0 +1,494 @@
...
+the scale of your application.
+
+:::
+
+Wait for the virtual machine (VM) to start up and copy the IP address.
+
+This guide uses `oauth2.example.com` as a hostname to run Ory Hydra. Replace it with `<something>.<your-domain>` for your
+purposes. You need to configure DNS, create an `A type` record, and point it to the VM's IP.
```
[the remark this answers is not in the document]
Comment (tomekpapiernik) [reply docs/hydra/guides/deploy-hydra-example.mdx:44]: @vinckr

File: docs/hydra/guides/deploy-hydra-example.mdx (hunk 5 above)
[the remark this answers is not in the document]
Comment (vinckr) [reply docs/hydra/guides/deploy-hydra-example.mdx:46]: title case? 
we should really document this as a convention in our styleguide

File: docs/hydra/guides/deploy-hydra-example.mdx [hunk 10]
```diff
@@ -0,0 +1,494 @@
...
+# Change the default password encryption to stronger one
+SELECT pg_reload_conf();
+# Reload configuration
+```
+
+Let's create a user for Hydra (Use your own password / hash!):
+
+```bash
+CREATE USER hydra PASSWORD 'CHANGE-ME-INSECURE-PASSWORD';
+```
```
[the remark this answers is not in the document]
Comment (vinckr) [reply docs/hydra/guides/deploy-hydra-example.mdx:105]: we should document some convention how we write these things <YOUR_PASSWORD_HERE> or 'CHANGE-ME-INSECURE-PASSWORD' are both fine IMO, would just be good to use something consistently.

File: docs/hydra/guides/deploy-hydra-example.mdx (hunk 3 above)
Comment (vinckr) [reply docs/hydra/guides/deploy-hydra-example.mdx:166]: but then the 
```
${useLatestRelease(
  "hydra"
)}
```
hook doesnt work?

Comment (tomekpapiernik) [reply docs/hydra/guides/deploy-hydra-example.mdx]: ah, that's a good point

**Review by vinckr — DISMISSED** [review]
LGMT 😛

**Pull request merged**